lib/Dialect/FIRRTL/Transforms/Dedup.cpp

```diff
@@ -1287,6 +1294,14 @@ class DedupPass : public DedupBase<DedupPass> {
       auto it = moduleHashes.find(h);
       if (it != moduleHashes.end()) {
         auto original = cast<FModuleLike>(it->second);
+        // If the original module comes after the current module, then dedup the
+        // original into the current module..
+        if (declarationOrder[original] > declarationOrder[module]) {
+          deduper.record(module);
+          dedupMap[moduleName] = moduleName;
```

Comment:
AFAICT this line is redundant with the later `dedupMap[moduleName] = original.moduleNameAttr()` below?  Just checking this is doing what is expected.

More importantly...
How does this work when there are multiple modules in the same (hash-)equivalence class-- if it's important that dedupMap's key/value pairs have value pointing to the EC leader everything is dedup'ed into ("dedup group" language suggests so, still reading the code), then won't chains of these end up possibly pointing to one another and not the final living module?

Reply:
Looking into this a bit... adding this to `dedup.mlir` causes a crash, I think because of the above and the code that follows validating MustDedup annotations...  

```mlir
// CHECK-LABEL: firrtl.circuit "DeclarationOrderNotInstantiationOrder2"
firrtl.circuit "DeclarationOrderNotInstantiationOrder2" attributes {annotations = [{
    // The annotation should be removed.
    // CHECK-NOT: class = "firrtl.transforms.MustDeduplicateAnnotation"
    class = "firrtl.transforms.MustDeduplicateAnnotation",
    modules = [
      "~DeclarationOrderNotInstantiationOrder2|X",
      "~DeclarationOrderNotInstantiationOrder2|Y",
      "~DeclarationOrderNotInstantiationOrder2|Z"
    ]}]
   } {
  // CHECK-NEXT: firrtl.module private @X
  firrtl.module private @X(in %a: !firrtl.uint<1>) {}
  firrtl.module private @Y(in %a: !firrtl.uint<1>) {}
  firrtl.module private @Z(in %a: !firrtl.uint<1>) {}
  firrtl.module @DeclarationOrderNotInstantiationOrder2() {
    %z_a = firrtl.instance y @Z(in a: !firrtl.uint<1>)
    %y_a = firrtl.instance y @Y(in a: !firrtl.uint<1>)
    %x_a = firrtl.instance x @X(in a: !firrtl.uint<1>)
  }
}
```

Reply:
Yeah, was going to say the same thing.  Not elegantly put but:
The `dedupMap` assumes that every deduped module A points to the thing it was deduped with B.  If B is later deduped with something else C, we need to update A to point to C or modify `getLead` to keep grabbing the lead until it finds the real head (`dedupMap[moduleName] == moduleName`)

Reply:
This is what I get for trying to fix this super quickly without reading the code. 😂

Reply:
Actually reading the code... solving this will be tricky. It would be simpler if the equivalence classes could be built and then a new leader (the winning deduped module) assigned based on the declaration order. However, this seems to run counter to how the pass is eagerly updating based on the first module found being assumed to be the leader. 🤔 I'm going to just close this and leave it for someone else as this PR was clearly half-baked. Thanks for keeping bad code in check. 😅